### PR TITLE
feat: modal form for reporting inaccurate Planning Constraints

### PR DESCRIPTION
**Pull request by jessicamcinchak**
**Changes:**
- Adds a new feature flag `"OVERRIDE_CONSTRAINTS"` so we can work on followup PRs in parallel
- Adds a link & modal to capture user input about inaccurate constraints (on intersecting constraints _only_):
  - The modal has context text, a checklist of entities that we think apply, and a long text input to provide a reason why it's inaccurate
  - "Submit" button does validation and updates internal component state with valid input
  - "Cancel" button clears input and exits the modal
- Basic styling feedback if you've marked an entity as inaccurate: gray text and "[Not applicable]" tag next to entity, and overall gray text accordion title if _every_ entity within that category has been marked as inaccurate

**Followup PRs:**
- Styling questions for Ian:
  - There's a mix of internal & external links in each accordion drawer, how best to distinguish?
  - More polished visual feedback after marking a constraint as inaccurate
  - Modal accessibility (currently re-using a lot of configs of UploadAndLabel modal)
- Passport logic for overall component `onSubmit` & automations 
- TESTS !

**Testing thread here:** https://opensystemslab.slack.com/archives/C04DZ1NBUMR/p1721219483725279

### CODE REVIEW

**Comment by jessicamcinchak** [comment]
[Ability to override planning constraints [Solution to GIS layers pulling constraints (and planning histories) that apply to adjoining properties]](https://trello.com/c/S6zRlO42/2637-ability-to-override-planning-constraints-solution-to-gis-layers-pulling-constraints-and-planning-histories-that-apply-to-adjoini)

File: editor.planx.uk/src/@planx/components/PlanningConstraints/Modal.tsx [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+    } else {
+      // Update inaccurateConstraints in parent component state on valid submit
+      const newInaccurateConstraints = {
+        ...inaccurateConstraints,
+        ...{ [fn]: { entities: checkedOptions, reason: textInput } },
+      };
+      setInaccurateConstraints(newInaccurateConstraints);
+      setShowModal(false);
+    }
+  };
```
Comment (jessicamcinchak) [inline editor.planx.uk/src/@planx/components/PlanningConstraints/Modal.tsx:120]: I've written my own simple validation here to start (similar to `UploadAndLabel` modal), but perhaps a good case to formalise using Formik/yup later down the line.

File: editor.planx.uk/src/@planx/components/PlanningConstraints/Public.tsx [hunk 2]
```diff
@@ -154,6 +166,8 @@ function Component(props: Props) {
               if (data) _constraints.push(data as GISResponse["constraints"]);
             }
 
+            const _overrides = inaccurateConstraints;
```
Comment (jessicamcinchak) [inline editor.planx.uk/src/@planx/components/PlanningConstraints/Public.tsx:169]: Simply writing a new raw `_overrides` key to the passport for now, follow-up PR will work out which values need to be added/removed from `property.constraints.planning` & `_nots` based on override entities and test automations.

File: editor.planx.uk/src/@planx/components/PlanningConstraints/Modal.tsx (hunk 1 above)
Comment (DafyddLlyr) [reply editor.planx.uk/src/@planx/components/PlanningConstraints/Modal.tsx:120]: Ah just seen a11y as a follow up listen in PR description sorry!

File: editor.planx.uk/src/@planx/components/PlanningConstraints/Modal.tsx [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+      }
+
+      if (newCheckedIds.length > 0) {
+        setShowChecklistError(false);
+      }
+      setCheckedOptions(newCheckedIds);
+    };
+
+  const changeInput = (e: React.ChangeEvent<HTMLInputElement>) => {
+    if (e.target.value.length > 1) {
```
Comment (DafyddLlyr) [inline editor.planx.uk/src/@planx/components/PlanningConstraints/Modal.tsx:93]:
```suggestion
    if (e.target.value.length) {
```

Little bug here - currently the error state persists if there's a single character submitted, once the error state is already triggered

image

File: editor.planx.uk/src/@planx/components/PlanningConstraints/Modal.tsx [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+  const closeModal = (_event: any, reason?: string) => {
+    if (reason && reason == "backdropClick") {
+      return;
+    }
+
+    // Clear any non-submitted inputs on cancel & sync parent state
+    setCheckedOptions(undefined);
+    setTextInput(undefined);
+    const newInaccurateConstraints = omit(inaccurateConstraints, fn);
+    setInaccurateConstraints(newInaccurateConstraints);
```
Comment (DafyddLlyr) [inline editor.planx.uk/src/@planx/components/PlanningConstraints/Modal.tsx:69]: This was a bit unexpected for a "cancel" interaction - I'd expect the initial state to persist, just changes I've made to be lost on cancel. This is bit more like a reset.

File: editor.planx.uk/src/@planx/components/PlanningConstraints/Modal.tsx [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+    };
+
+  const changeInput = (e: React.ChangeEvent<HTMLInputElement>) => {
+    if (e.target.value.length > 1) {
+      setShowInputError(false);
+    }
+    setTextInput(e.target.value);
+  };
+
+  const validateAndSubmit = () => {
```
Comment (DafyddLlyr) [inline editor.planx.uk/src/@planx/components/PlanningConstraints/Modal.tsx:99]: One edge-case workflow - 

- I mark a constraint as not applying
- I re-open modal to uncheck
- I can't save without error
- I have to cancel to get back to no overrides

This ties into the comment above on cancel behaviour - I think it is valid to submit this form with nothing checked, if the form is dirty.

**Review by DafyddLlyr — CHANGES_REQUESTED** [review]
A few minor comments and suggestions 🙂

File: editor.planx.uk/src/@planx/components/PlanningConstraints/Modal.tsx (hunk 1 above)
Comment (jessicamcinchak) [reply editor.planx.uk/src/@planx/components/PlanningConstraints/Modal.tsx:130]: Updated outermost modal `Box` to be `component="form"` and added `type="submit"` to button :+1: 

Agree Formik is still one to come back to!

File: editor.planx.uk/src/@planx/components/PlanningConstraints/List.tsx [hunk 6]
```diff
@@ -131,21 +143,27 @@ export default function ConstraintsList({
...
+  fn: Constraint["fn"];
   value: Constraint["value"];
   content: Constraint["text"];
   data: Constraint["data"] | null;
   metadata?: Metadata;
   category: string;
   children: ReactNode;
+  inaccurateConstraints: InaccurateConstraints;
+  setInaccurateConstraints: (
+    value: React.SetStateAction<InaccurateConstraints>,
```
[the remark this answers is not in the document]
Comment (jessicamcinchak) [reply editor.planx.uk/src/@planx/components/PlanningConstraints/List.tsx:155]: Yep will make a note of this in nice-to-haves / future complexity thing !

File: editor.planx.uk/src/@planx/components/PlanningConstraints/Modal.tsx [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+        newCheckedIds = checkedOptions.filter((e) => e !== id);
+      } else {
+        newCheckedIds = [...(checkedOptions || []), id];
+      }
+
+      if (newCheckedIds.length > 0) {
+        setShowChecklistError(false);
+      }
+      setCheckedOptions(newCheckedIds);
+    };
```
[the remark this answers is not in the document]
Comment (jessicamcinchak) [reply editor.planx.uk/src/@planx/components/PlanningConstraints/Modal.tsx:91]: Fully agree and went through a similar process before also eventually landing on this checkbox solution. I basically had a half-formik controlled modal on first draft (controlling the text input but not checklist) but was sinking time, so ended up with custom validation for now.

Feels like a nice one to pick up independently later once component tests are in place here and we can tackle independent of feature logic :+1:

File: editor.planx.uk/src/@planx/components/PlanningConstraints/Modal.tsx [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+              </ErrorWrapper>
+            </InputLabel>
+          </Box>
+          <Box>
+            <InputLabel label="Tell us why" htmlFor="reason">
+              <ErrorWrapper
+                error={showInputError ? ERROR_MESSAGES["input"] : undefined}
+                id={`input-error-inaccurate-entities`}
+              >
+                <Input
```
[the remark this answers is not in the document]
Comment (jessicamcinchak) [reply editor.planx.uk/src/@planx/components/PlanningConstraints/Modal.tsx:213]: Made a `todo` note for this one - multiple errors is yet another good reason to come back to formik or at minimum a more formalised validation schema :+1:

File: editor.planx.uk/src/@planx/components/PlanningConstraints/Modal.tsx (hunk 3 above)
Comment (jessicamcinchak) [reply editor.planx.uk/src/@planx/components/PlanningConstraints/Modal.tsx]: Now resolved!

File: editor.planx.uk/src/@planx/components/PlanningConstraints/Modal.tsx (hunk 4 above)
Comment (jessicamcinchak) [reply editor.planx.uk/src/@planx/components/PlanningConstraints/Modal.tsx]: Good catch - I was treating this like a "clear/reset". Now adjusted to reset to initial component state on "Cancel" :+1:

File: editor.planx.uk/src/@planx/components/PlanningConstraints/Modal.tsx (hunk 5 above)
Comment (jessicamcinchak) [reply editor.planx.uk/src/@planx/components/PlanningConstraints/Modal.tsx:100]: Can recreate and think I've got a solution for this one: 
- if you open an empty form and click "Submit", I'll still trigger the validation error messages and prevent submission
- if you open a pre-populated form, clear your selected options and text input, and click "Submit" - I'll sync your updates to parent state (essentially "clear") and close the modal
  - is it obvious you have to manually clear _both_ your selected options and text input? Would you expect text input to auto-clear as soon as zero checklist options are selected?

Comment (DafyddLlyr) [reply editor.planx.uk/src/@planx/components/PlanningConstraints/Modal.tsx:100]: Working as expected!

One very minor issue is that errors persist from the modal being closed and opened - I guess they should be cleared on both submit and cancel?

**Review by DafyddLlyr — APPROVED** [review]

**Pull request merged**